api/src/main/java/org/openmrs/module/eptsreports/reporting/library/datasets/ListChildrenOnARTandFormulationsDataset.java

```diff
@@ -1,24 +1,784 @@
+            + "             GROUP BY p.patient_id "
+            + "                   ) most_recent  ON p.patient_id = most_recent.patient_id   "
+            + " WHERE  p.voided = 0  "
+            + "     AND e.voided = 0  "
+            + "     AND o.voided = 0  "
+            + "     AND e.location_id = :location "
+            + "     AND e.encounter_type IN (${18}) "
+            + "     AND e.encounter_datetime <= :endDate "
+            + "     AND e.encounter_datetime = most_recent.e_datetime "
+            + "     AND o.concept_id = ${5096} ";
+
+    StringSubstitutor substitutor = new StringSubstitutor(valuesMap);
+
+    spdd.setQuery(substitutor.replace(sql));
+    return spdd;
+  }
+
+  /**
+   * 14
+   *
+   * @return
+   */
+  private DataDefinition getLastFollowupConsultationDate() {
+    SqlPatientDataDefinition spdd = new SqlPatientDataDefinition();
+    spdd.setName("Last Follow up Consultation Date");
+    spdd.addParameter(new Parameter("location", "location", Location.class));
+    spdd.addParameter(new Parameter("endDate", "endDate", Date.class));
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("6", hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("9", hivMetadata.getPediatriaSeguimentoEncounterType().getEncounterTypeId());
+
+    String sql =
+        " SELECT p.patient_id, MAX(e.encounter_datetime) "
+            + " FROM   patient p  "
+            + "         INNER JOIN encounter e  "
+            + "                         ON p.patient_id = e.patient_id  "
+            + "         INNER JOIN obs o  "
+            + "                         ON e.encounter_id = o.encounter_id  "
+            + "WHERE  p.voided = 0  "
+            + "         AND e.voided = 0  "
+            + "         AND o.voided = 0  "
+            + "         AND e.location_id = :location "
+            + "         AND e.encounter_type IN (${6}, ${9}) "
+            + "         AND e.encounter_datetime <= :endDate "
+            + "GROUP BY p.patient_id ";
+
+    StringSubstitutor substitutor = new StringSubstitutor(valuesMap);
+
+    spdd.setQuery(substitutor.replace(sql));
+    return spdd;
+  }
+
+  /**
+   * 15
+   *
+   * @return
+   */
+  private DataDefinition getARVRegimenLastConsultationDate() {
+    SqlPatientDataDefinition spdd = new SqlPatientDataDefinition();
+    spdd.setName("ARV Regimen on Last Consultation");
+    spdd.addParameter(new Parameter("location", "location", Location.class));
+    spdd.addParameter(new Parameter("endDate", "endDate", Date.class));
+
+    Map<String, Integer> valuesMap = new HashMap<>();
+    valuesMap.put("6", hivMetadata.getAdultoSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("9", hivMetadata.getPediatriaSeguimentoEncounterType().getEncounterTypeId());
+    valuesMap.put("1087", hivMetadata.getPreviousARVUsedForTreatmentConcept().getConceptId());
+    String sql =
+        "  SELECT p.patient_id, o.value_coded "
```

Comment:
same idea here, the name will make the report more human readable

Reply:
done